File: app/src/main/java/io/github/freewebmovement/igniter/activities/HomeFragment.java
package io.github.freewebmovement.igniter.activities;

import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import io.github.freewebmovement.igniter.IgniterApplication;
import io.github.freewebmovement.igniter.R;
import io.github.freewebmovement.igniter.services.ProxyService;

/**
 * Connect page: current state, a big start/stop toggle, the active server card
 * and a connection test.
 */
public class HomeFragment extends Fragment {

    private TextView mStatusTitle;
    private TextView mStatusSub;
    private TextView mServerValue;
    private Button mConnectBtn;

    public static HomeFragment newInstance() {
        return new HomeFragment();
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_home, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mStatusTitle = view.findViewById(R.id.homeStatusTitle);
        mStatusSub = view.findViewById(R.id.homeStatusSub);
        mServerValue = view.findViewById(R.id.homeServerValue);
        mConnectBtn = view.findViewById(R.id.homeConnectBtn);

        mConnectBtn.setOnClickListener(v -> {
            MainActivity activity = (MainActivity) getActivity();
            if (activity == null) {
                return;
            }
            if (activity.isProxyRunning()) {
                activity.stopProxy();
            } else {
                activity.startProxy();
            }
        });
        view.findViewById(R.id.homeServerRow).setOnClickListener(v -> {
            MainActivity activity = (MainActivity) getActivity();
            if (activity != null) {
                activity.openServersTab();
            }
        });
        view.findViewById(R.id.homeTestBtn).setOnClickListener(v -> {
            MainActivity activity = (MainActivity) getActivity();
            if (activity != null) {
                activity.testConnection();
            }
        });

        refreshServerInfo();
        updateState(((MainActivity) requireActivity()).getProxyState());
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        inflater.inflate(R.menu.menu_home, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        MainActivity activity = (MainActivity) getActivity();
        if (activity == null) {
            return false;
        }
        if (id == R.id.action_view_test_connection) {
            activity.testConnection();
            return true;
        }
        if (id == R.id.action_view_settings) {
            startActivity(new Intent(activity, SettingsActivity.class));
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onResume() {
        super.onResume();
        refreshServerInfo();
    }

    public void updateState(int state) {
        if (mStatusTitle == null || mConnectBtn == null) {
            return;
        }
        switch (state) {
            case ProxyService.STARTED: {
                mStatusTitle.setText(R.string.home_status_started);
                mStatusTitle.setTextColor(0xFF2E7D32);
                mConnectBtn.setText(R.string.home_btn_stop);
                mConnectBtn.setEnabled(true);
                break;
            }
            case ProxyService.STARTING:
            case ProxyService.STOPPING: {
                mStatusTitle.setText(state == ProxyService.STARTING
                        ? R.string.home_status_starting : R.string.home_status_stopping);
                mStatusTitle.setTextColor(0xFFF57F17);
                mConnectBtn.setEnabled(false);
                break;
            }
            default: {
                mStatusTitle.setText(R.string.home_status_stopped);
                mStatusTitle.setTextColor(0xFF757575);
                mConnectBtn.setText(R.string.home_btn_start);
                mConnectBtn.setEnabled(true);
                break;
            }
        }
    }

    public void refreshServerInfo() {
        if (mServerValue == null || mStatusSub == null) {
            return;
        }
        IgniterApplication app = IgniterApplication.getApplication();
        String addr = app.trojanConfig.getRemoteAddr();
        int port = app.trojanConfig.getRemotePort();
        if (addr == null || addr.isEmpty()) {
            mServerValue.setText(R.string.home_server_unknown);
            mStatusSub.setText(R.string.home_server_unknown);
        } else {
            String text = addr + ":" + port;
            mServerValue.setText(text);
            mStatusSub.setText(text);
        }
    }
}

File: app/src/main/java/io/github/freewebmovement/igniter/activities/HomeFragment.kt
package io.github.freewebmovement.igniter.activities

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.platform.ViewCompositionStrategy
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.fragment.app.viewModels
import io.github.freewebmovement.igniter.IgniterApplication
import io.github.freewebmovement.igniter.R
import io.github.freewebmovement.igniter.activities.servers.activity.AddServerActivity
import io.github.freewebmovement.igniter.activities.servers.viewmodel.ServerListViewModel
import io.github.freewebmovement.igniter.connection.ServerPingManager
import io.github.freewebmovement.igniter.persistence.TrojanConfig
import io.github.freewebmovement.igniter.persistence.database.Server
import io.github.freewebmovement.igniter.services.ProxyService
import io.github.freewebmovement.igniter.services.ServerPingService
import io.github.freewebmovement.igniter.theme.IgniterTheme
import io.github.freewebmovement.igniter.ui.home.HomeTopBar

/**
 * Proxy servers page: current state, a big start/stop toggle, a connection test,
 * the proxy route summary and the server list with add/select/delete/import.
 * UI is Jetpack Compose via [HomeTopBar].
 */
class HomeFragment : Fragment() {

    private val homeViewModel: HomeViewModel by activityViewModels()
    private val serverViewModel: ServerListViewModel by viewModels()

    private var mProxyState by mutableStateOf(ProxyService.STATE_NONE)
    private var mServerSummary by mutableStateOf("")
    private var mRouteSummary by mutableStateOf("")
    private var mDomainRouteCount by mutableStateOf(0)
    private var mServers by mutableStateOf<List<Server>>(emptyList())
    private var mPingData by mutableStateOf<Map<String, ServerPingManager.PingInfo>>(emptyMap())
    private var mTestResults by mutableStateOf<Map<String, Pair<Boolean, Long>>>(emptyMap())
    private var mCurrentHost by mutableStateOf("")
    private var mCurrentPort by mutableStateOf(0)
    private var mPendingStartAfterStop = false

    private val addServerLauncher = registerForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) { result ->
        if (result.resultCode == Activity.RESULT_OK) {
            Toast.makeText(requireContext(), R.string.add_server_success, Toast.LENGTH_SHORT).show()
        }
    }

    private val editServerLauncher = registerForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) { result ->
        if (result.resultCode == Activity.RESULT_OK) {
            Toast.makeText(requireContext(), R.string.common_save_success, Toast.LENGTH_SHORT).show()
        }
    }

    private val fileImportLauncher = registerForActivityResult(
        ActivityResultContracts.GetContent()
    ) { uri ->
        if (uri != null) {
            serverViewModel.importConfigsFromFile(uri)
            Toast.makeText(requireContext(), R.string.add_server_success, Toast.LENGTH_SHORT).show()
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View {
        return ComposeView(requireContext()).apply {
            setViewCompositionStrategy(ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed)
            setContent {
                IgniterTheme {
                    HomeTopBar(
                        proxyState = mProxyState,
                        serverSummary = mServerSummary,
                        routeSummary = mRouteSummary,
                        domainRouteCount = mDomainRouteCount,
                        servers = mServers,
                        pingData = mPingData,
                        testResults = mTestResults,
                        currentHost = mCurrentHost,
                        currentPort = mCurrentPort,
                        onImportFromFile = { fileImportLauncher.launch("text/plain") },
                        onConnectClick = {
                            val act = activity as? MainActivity ?: return@HomeTopBar
                            val state = act.getProxyState()
                            if (state == ProxyService.STARTED || state == ProxyService.STARTING) {
                                act.stopProxy()
                                homeViewModel.setTestingServer(null)
                                val key = "${mCurrentHost}:${mCurrentPort}"
                                mTestResults = mTestResults - key
                            } else {
                                val host = mCurrentHost
                                val port = mCurrentPort
                                if (host.isNotEmpty()) {
                                    homeViewModel.setTestingServer("$host:$port")
                                }
                                act.startProxy()
                            }
                        },
                        onDomainRouteClick = { (activity as? MainActivity)?.openRulesTab() },
                        onAddServerClick = {
                            addServerLauncher.launch(
                                Intent(requireContext(), AddServerActivity::class.java))
                        },
                        onServerSelected = { server ->
                            (activity as? MainActivity)?.onServerSelected(toTrojanConfig(server))
                            refreshCurrentServer()
                        },
                        onServerPlay = { server ->
                            val act = activity as? MainActivity ?: return@HomeTopBar
                            act.onServerSelected(toTrojanConfig(server))
                            homeViewModel.setTestingServer("${server.hostname}:${server.port}")
                            val state = act.getProxyState()
                            if (state == ProxyService.STARTED || state == ProxyService.STARTING) {
                                mPendingStartAfterStop = true
                                act.stopProxy()
                            } else {
                                act.startProxy()
                            }
                        },
                        onServerStop = { server ->
                            val act = activity as? MainActivity ?: return@HomeTopBar
                            val config = IgniterApplication.getApplication().trojanConfig
                            if (server.hostname == config.getRemoteAddr() &&
                                server.port == config.getRemotePort()
                            ) {
                                act.stopProxy()
                                homeViewModel.setTestingServer(null)
                                mTestResults = mTestResults - "${server.hostname}:${server.port}"
                            }
                        },
                        onServerTest = { _ ->
                            (activity as? MainActivity)?.testConnection()
                        },
                        onServerEdit = { server ->
                            val intent = Intent(requireContext(), AddServerActivity::class.java)
                            intent.putExtra(AddServerActivity.EXTRA_SERVER_ID, server.id)
                            editServerLauncher.launch(intent)
                        },
                        onServerDelete = { server ->
                            val act = activity as? MainActivity
                            if (act != null) {
                                val config = IgniterApplication.getApplication().trojanConfig
                                if (server.hostname == config.getRemoteAddr() &&
                                    server.port == config.getRemotePort()
                                ) {
                                    act.stopProxy()
                                }
                            }
                            serverViewModel.deleteServer(server)
                        }
                    )
                }
            }
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        homeViewModel.proxyState.observe(viewLifecycleOwner) {
            mProxyState = it
            if (it == ProxyService.STARTED) {
                view?.postDelayed({ (activity as? MainActivity)?.testConnection() }, TEST_DELAY_MS)
            }
            if (it == ProxyService.STOPPED && mPendingStartAfterStop) {
                mPendingStartAfterStop = false
                (activity as? MainActivity)?.startProxy()
            }
            refreshCurrentServer()
        }
        homeViewModel.serverSummary.observe(viewLifecycleOwner) { mServerSummary = it }
        homeViewModel.routeSummary.observe(viewLifecycleOwner) { mRouteSummary = it }
        homeViewModel.domainRouteCount.observe(viewLifecycleOwner) { mDomainRouteCount = it }
        homeViewModel.refreshServer()
        homeViewModel.refreshRoute()

        serverViewModel.servers.observe(viewLifecycleOwner) { servers ->
            mServers = servers
            refreshCurrentServer()
        }

        ServerPingManager.pingData.observe(viewLifecycleOwner) { mPingData = it }

        homeViewModel.testResult.observe(viewLifecycleOwner) { result ->
            if (result != null) {
                mTestResults = mTestResults + (result.serverKey to Pair(result.connected, result.delay))
            }
        }

        refreshCurrentServer()
    }

    private fun refreshCurrentServer() {
        val config = IgniterApplication.getApplication().trojanConfig
        mCurrentHost = config.getRemoteAddr()
        mCurrentPort = config.getRemotePort()
    }

    private fun toTrojanConfig(server: Server): TrojanConfig {
        val config = TrojanConfig()
        config.setRemoteAddr(server.hostname)
        config.setRemotePort(server.port)
        config.setPassword(server.password)
        config.setLocalPort(server.local_port)
        return config
    }

    override fun onResume() {
        super.onResume()
        homeViewModel.refreshServer()
        homeViewModel.refreshRoute()
        ServerPingService.start(requireContext())
    }

    override fun onPause() {
        super.onPause()
        ServerPingService.stop(requireContext())
    }

    override fun onDestroyView() {
        super.onDestroyView()
    }

    companion object {
        private const val TEST_DELAY_MS = 3000L

        @JvmStatic
        fun newInstance(): HomeFragment {
            return HomeFragment()
        }
    }
}
